### PR TITLE
perf: reduce key cloning in SsTableBuilder::add_inner

### PR DESCRIPTION
**Pull request by ben1009**
## Summary

Remove live `KeyVec` clones from `BlockBuilder` and `SsTableBuilder`. Block-meta first/last keys are now derived from the already-encoded block data via a new `key_at()` method, eliminating 3 key copies per `add_inner` call.

### Core Optimization
- Remove `first_key: KeyVec` from `BlockBuilder`; store offset+len instead
- Add `key_at(idx)` to reconstruct any key from encoded block entries without live clones
- Add `has_first_key` sentinel to correctly handle empty first keys
- Derive `SsTableBuilder` meta keys from `key_at` at block boundaries only
- Move `shared_bytes_from_slice` to `key.rs`, removing duplication between `mem_table.rs` and `block/builder.rs`

### Correctness Improvements
- Replace silent `let _ =` ignore on oversized entry with `bail!` (defensive safety net)
- Add `has_data` sentinel for `SsTableBuilder::is_empty()` (fixes false-empty on actual empty first keys)

### Tests
- `key_at`: multi-entry, single entry, empty first key, long keys (>128B chunked path), out-of-bounds panic
- `SsTableBuilder::is_empty()`: fresh builder, after add, after block seal+re-add
- Oversized entry `bail!` path with tiny block_size

## Benchmark Results

Criterion back-to-back comparison (baseline = pre-change):

| Workload | Wall-Clock Change |
|----------|-------------------|
| `write_throughput/inline/4KB`  | **−1.0%** |
| `write_throughput/inline/16KB` | +1.9% (noise) |
| `write_throughput/inline/64KB` | **−1.4%** |
| `compaction/inline`            | **−7.0%** (p=0.00) |
| `flush_throughput/inline`      | +8.4% (noise) |

## CPU Profile

`SsTableBuilder::add_inner` remains the dominant hotspot at ~45% of cycles. Hardware counters (IPC, cache misses, branch misses) are within noise vs baseline.

## Verification
- ✅ `cargo fmt --check` passes
- ✅ `cargo clippy` passes
- ✅ 143 tests pass
- ✅ Criterion benchmarks show no regression

## Summary by CodeRabbit

* **Performance Improvements**
  * ~3–5% write CPU improvement from reduced key cloning and leaner block write handling.

* **Bug Fixes**
  * Oversized entries now produce detailed errors instead of being silently ignored.
  * More robust first/last-key derivation and entry boundary handling to improve correctness.

* **Documentation**
  * Updated performance profile with optimization details and post-change CPU snapshots.

* **Tests**
  * Added unit tests for key reconstruction, block edge cases, and builder seal/re-add paths.

* **Benchmarks**
  * Adjusted write-throughput benchmark value-size cap for accurate inline-size limits.

### CODE REVIEW

File: kv-engine/src/block/builder.rs [hunk 1]
```diff
@@ -34,65 +41,97 @@ impl BlockBuilder {
...
         self.data.put(value);
 
-        if self.first_key.is_empty() {
-            self.first_key = key.to_key_vec();
+        if !self.has_first_key {
+            // First entry always has overlap == 0, so the suffix is the full key.
+            self.has_first_key = true;
+            self.first_key_offset = entry_start + 2 * SIZE_OF_U16;
+            self.first_key_len = key.len() as u16;
         }
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/block/builder.rs]: Fixed in commit 6fe3365. All `as u16` casts replaced with `u16::try_from` + `map_err`. `BlockBuilder::add()` now returns `Result<bool>` and propagates errors via `?`.

File: kv-engine/src/block/builder.rs [hunk 2]
```diff
@@ -34,65 +42,115 @@ impl BlockBuilder {
...
+    ///
+    /// Returns `Err` if any field exceeds `u16::MAX` (the block format stores
+    /// lengths and offsets as `u16`).
+    pub fn add(&mut self, key: KeySlice, value: &[u8]) -> Result<bool> {
+        // The first entry always fits; skip the size check.
         if !self.is_empty()
             && self.current_size() + key.len() + value.len() + 3 * SIZE_OF_U16 >= self.block_size
         {
-            return false;
+            return Ok(false);
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/block/builder.rs:73]: Not fixed — intentional design. The "first entry always fits" rule is required for correctness: `add_inner` in `SsTableBuilder` seals a block when `add()` returns `false`, then retries on a fresh builder. If the fresh builder also rejected its first entry, the `bail!` path would fire for normal-sized entries that simply exceed the remaining space in the previous block. Removing this special case would break block sealing. The `bail!` at line 144 of `table/builder.rs` is a defensive safety net that is unreachable in correct code.

File: kv-engine/src/block/builder.rs [hunk 3]
```diff
@@ -34,65 +42,115 @@ impl BlockBuilder {
...
         self.data.put(value);
 
-        if self.first_key.is_empty() {
-            self.first_key = key.to_key_vec();
+        if !self.has_first_key {
+            // First entry always has overlap == 0, so the suffix is the full key.
+            self.has_first_key = true;
+            self.first_key_offset = entry_start + 2 * SIZE_OF_U16;
+            self.first_key_len = u16::try_from(key.len())
+                .map_err(|_| anyhow::format_err!("key length {} exceeds u16::MAX", key.len()))?;
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/block/builder.rs]: Fixed in commit 6fe3365. All `u16::try_from` conversions now happen before any mutation of `offsets` or `data`. Values are validated first, then written atomically.

**Comment by ben1009** [comment]
All review comments have been addressed and resolved. Please re-review.

**Summary of fixes:**
1. **Gemini** (u16 truncation): All `as u16` casts replaced with `u16::try_from` + error propagation. `BlockBuilder::add()` now returns `Result<bool>`.
2. **CodeRabbit** (oversized first entry): Not fixed — intentional design. The first-entry-accepts rule is required for block sealing correctness in `SsTableBuilder::add_inner`.
3. **CodeRabbit** (transactional add): All u16 conversions now happen before any mutation of `offsets` or `data`.

@gemini-code-assist @coderabbitai

File: kv-engine/src/block/builder.rs [hunk 4]
```diff
@@ -1,7 +1,8 @@
+use anyhow::Result;
 use bytes::{BufMut, Bytes};
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/block/builder.rs:2]: Not fixed. The `bytes::Buf` trait is not needed here — `key_at` reads from `self.data` (a `Vec<u8>`) using direct indexing, which is clear and correct. Adding `Buf` would be an unused import.

File: kv-engine/src/block/builder.rs [hunk 5]
```diff
@@ -34,65 +42,115 @@ impl BlockBuilder {
...
+        assert!(idx < self.num_entries(), "key_at index out of bounds");
+
+        let off = self.offsets[idx] as usize;
+        let overlap = u16::from_be_bytes([self.data[off], self.data[off + 1]]) as usize;
+        let suffix_len = u16::from_be_bytes([
+            self.data[off + SIZE_OF_U16],
+            self.data[off + SIZE_OF_U16 + 1],
+        ]) as usize;
+        let suffix_start = off + 2 * SIZE_OF_U16;
+        let suffix = &self.data[suffix_start..suffix_start + suffix_len];
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/block/builder.rs:129]: Not fixed. The block format stores u16 fields in big-endian (the writer uses `put_u16` which is big-endian). Using `get_u16_le()` as suggested would read little-endian and corrupt data. The current `u16::from_be_bytes` is correct and consistent with the encoding.
